rust/diesel/src/order.rs

```diff
@@ -0,0 +1,173 @@
+use std::ops::{Add, Mul};
+
+use crate::product::Product;
+use crate::schema::order_lines;
+use crate::schema::orders;
+use num_traits::identities::Zero;
+
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct Order {
+    pub order_id: i32,
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "orders"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+struct NewOrder {
+    pub order_total: BigDecimal,
+    pub user_id: i32,
+}
+
+#[table_name = "order_lines"]
+#[derive(AsChangeset, Serialize, Deserialize, Queryable, Insertable)]
+#[serde(rename_all = "camelCase")]
+pub struct NewOrderLine {
+    pub order_id: i32,
+    pub product_id: i32,
+    pub units: i16,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NewUserOrder {
+    pub user_id: i32,
+    pub products: Vec<ProductOrder>,
+}
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ProductOrder {
+    pub product_id: i32,
+    pub units: i16,
+}
+
+impl Order {
+    pub fn create(order: NewUserOrder, connection: &PgConnection) -> Option<NewUserOrder> {
+        let product_units: Vec<(Product, i16)> = order
+            .products
+            .iter()
+            .flat_map(|product_ref| {
+                Product::read(product_ref.product_id, connection)
+                    .map(|product| (product, product_ref.units))
+            })
+            .collect();
+
+        let order_total = product_units
+            .to_vec()
+            .into_iter()
+            .fold(BigDecimal::from(0.0), |acc, (product, qty)| {
+                acc.add(product.price.mul(BigDecimal::from(qty)))
+            });
+
+        if order_total.is_zero() {
+            return None;
+        }
+
+        let row = NewOrder {
+            user_id: order.user_id,
+            order_total,
+        };
+
+        let insert_result = connection.build_transaction().read_write().run(|| {
+            let inserted_order: Order = diesel::insert_into(orders::table)
+                .values(&row)
+                .get_result(connection)
+                .unwrap();
+
+            let order_lines: Vec<NewOrderLine> = product_units
+                .to_vec()
+                .into_iter()
+                .map(|(product, units)| NewOrderLine {
+                    order_id: inserted_order.order_id,
+                    product_id: product.product_id,
+                    units,
+                })
+                .collect();
+
+            diesel::insert_into(order_lines::table)
+                .values(order_lines)
+                .execute(connection)
+        });
+
+        insert_result.map(|_| order).ok()
+    }
+
+    pub fn read_all(connection: &PgConnection) -> Vec<Order> {
+        orders::table
+            .order(orders::order_id)
+            .load::<Order>(connection)
+            .unwrap()
```

Comment:
I think the result should be propagated back to the caller, so it could be printed/handled. Not that we really need to handle it in this example app, but I think it is good practice.

Reply:
Again, in this example, it is sufficient to know if the operation failed, not why it failed, hence I will keep Option here